packages/vite/src/node/config.ts

```diff
@@ -49,7 +49,9 @@ export interface ConfigEnv {
   mode: string
 }
 
-export type UserConfigFn = (env: ConfigEnv) => UserConfig | Promise<UserConfig>
+export type UserConfigFn = (
+  env: Readonly<ConfigEnv>
+) => UserConfig | Promise<UserConfig>
```

Comment:
Maybe this change could be moved to another PR, as it is unrelated to the fix and it is easier to discuss it separately.

Reply:
My intention is to prevent people from mutating the object by type conversion to avoid undefined behavior. But you are right, it is unrelated to this PR. I will remove it.

Reply:
Yes, please create another PR with it if you want to discuss this change. It is just easier to discuss and get the approvals to merge them if the PRs are focused on only one thing.